### PR TITLE
interfaces: use ConnectedPlug/ConnectedSlot types (step 1)

### PR DESCRIPTION
**Pull request by stolowski**
This PR makes all *Connected* methods of interfaces use ConnectedPlug/Slot type. This is step 1 to limit the diff... There is a small temporary hack in *Connected* types to expose all attributes via Attrs public - just to reduce the changes here. The hack is going to be removed in the followup PR which completes the switch to ConnectedPlug/Slot API.

### CODE REVIEW

File: interfaces/builtin/all_test.go [hunk 1]
```diff
@@ -115,10 +115,10 @@ type seccompDefiner4 interface {
 }
 
 type systemdDefiner1 interface {
-	SystemdConnectedPlug(spec *systemd.Specification, plug *interfaces.Plug, plugAttrs map[string]interface{}, slot *interfaces.Slot, slotAttrs map[string]interface{}) error
```
Comment (zyga) [inline interfaces/builtin/all_test.go]: Can you please move the old definitions to "known old APIs" that we detect and error on?

**Comment by zyga** [comment]
There's a good chunk of test failures that look real.

**Comment by stolowski** [comment]
@zyga Ok, it was a simple omission that I already fixed in another branch, just forgot it's going to affect this PR too. Fixed now.

File: interfaces/builtin/all_test.go (hunk 1 above)
Comment (stolowski) [reply interfaces/builtin/all_test.go]: Done.

File: interfaces/repo.go [hunk 2]
```diff
@@ -579,8 +579,8 @@ func (r *Repository) Connect(ref ConnRef) error {
 	}
 
 	// TODO: store copy of attributes from hooks
-	cplug := &ConnectedPlug{plugInfo: plug}
-	cslot := &ConnectedSlot{slotInfo: slot}
+	cplug := NewConnectedPlug(plug, nil)
+	cslot := NewConnectedSlot(slot, nil)
```
Comment (niemeyer) [inline interfaces/repo.go]: Doesn't feel like that second argument is relevant. If you don't yet have a use case in other branches, can we please drop it?

Comment (stolowski) [reply interfaces/repo.go]: The second argument is for passing dynamic attributes when constructing connected plug/slot. It's not used in this branch but it is the core of the feature and is used in the followup integration branch.

File: interfaces/connection.go [hunk 3]
```diff
@@ -36,13 +36,17 @@ type ConnectedPlug struct {
 	plugInfo     *snap.PlugInfo
 	staticAttrs  map[string]interface{}
 	dynamicAttrs map[string]interface{}
+	// FIXME temporary
+	Attrs map[string]interface{}
```
[the remark this answers is not in the document]
Comment (stolowski) [reply interfaces/connection.go]: Yeah, #4314 gets rid of it and completes the transition.

File: interfaces/builtin/serial_port_test.go [hunk 4]
```diff
@@ -35,40 +36,70 @@ type SerialPortInterfaceSuite struct {
...
-	badPathSlot3     *interfaces.Slot
-	badPathSlot4     *interfaces.Slot
-	badPathSlot5     *interfaces.Slot
-	badPathSlot6     *interfaces.Slot
-	badPathSlot7     *interfaces.Slot
-	badPathSlot8     *interfaces.Slot
-	badPathSlot9     *interfaces.Slot
-	badPathSlot10    *interfaces.Slot
-	badInterfaceSlot *interfaces.Slot
+	testSlot1            *interfaces.ConnectedSlot
```
Comment (zyga) [inline interfaces/builtin/serial_port_test.go]: OMG

File: interfaces/builtin/serial_port_test.go [hunk 5]
```diff
@@ -134,25 +165,44 @@ slots:
...
-	s.badPathSlot3 = &interfaces.Slot{SlotInfo: osSnapInfo.Slots["bad-path-3"]}
-	s.badPathSlot4 = &interfaces.Slot{SlotInfo: osSnapInfo.Slots["bad-path-4"]}
-	s.badPathSlot5 = &interfaces.Slot{SlotInfo: osSnapInfo.Slots["bad-path-5"]}
-	s.badPathSlot6 = &interfaces.Slot{SlotInfo: osSnapInfo.Slots["bad-path-6"]}
-	s.badPathSlot7 = &interfaces.Slot{SlotInfo: osSnapInfo.Slots["bad-path-7"]}
-	s.badPathSlot8 = &interfaces.Slot{SlotInfo: osSnapInfo.Slots["bad-path-8"]}
-	s.badPathSlot9 = &interfaces.Slot{SlotInfo: osSnapInfo.Slots["bad-path-9"]}
-	s.badPathSlot10 = &interfaces.Slot{SlotInfo: osSnapInfo.Slots["bad-path-10"]}
-	s.badInterfaceSlot = &interfaces.Slot{SlotInfo: osSnapInfo.Slots["bad-interface"]}
+	s.testSlot1Info = osSnapInfo.Slots["test-port-1"]
```
Comment (zyga) [inline interfaces/builtin/serial_port_test.go]: OMG OMG

File: interfaces/builtin/spi.go [hunk 6]
```diff
@@ -56,10 +57,10 @@ func (iface *spiInterface) StaticInfo() interfaces.StaticInfo {
 
 var spiDevPattern = regexp.MustCompile("^/dev/spidev[0-9].[0-9]+$")
 
-func (iface *spiInterface) path(slot *interfaces.Slot) (string, error) {
-	path, ok := slot.Attrs["path"].(string)
+func (iface *spiInterface) path(attrs map[string]interface{}) (string, error) {
+	path, ok := attrs["path"].(string)
 	if !ok || path == "" {
-		return "", fmt.Errorf("slot %q must have a path attribute", slot.Ref())
+		return "", fmt.Errorf("spi slot must have a path attribute")
```
Comment (zyga) [inline interfaces/builtin/spi.go]: Hmmm, can we pass more data and retain  the error message

File: interfaces/builtin/utils_test.go [hunk 7]
```diff
@@ -70,3 +73,19 @@ func MockPlug(c *C, yaml string, si *snap.SideInfo, plugName string) *interfaces
 func MockSlot(c *C, yaml string, si *snap.SideInfo, slotName string) *interfaces.Slot {
 	return builtin.MockSlot(c, yaml, si, slotName)
 }
+
+func MockConnectedPlug(c *C, yaml string, si *snap.SideInfo, plugName string) (*interfaces.ConnectedPlug, *snap.PlugInfo) {
```
Comment (zyga) [inline interfaces/builtin/utils_test.go]: Can we reuse those from export test?

Comment (stolowski) [reply interfaces/builtin/utils_test.go]: No, not really, not easily at least. I once looked at it as we have other duplicated stuff there. They are in two separate packages (builtin and builtin_test) and we have tests depending on one or the other. Once I attempted to get rid of these duplicated helpers I got into a rabbit hole. Would be worth investigating at some point though.

**Pull request merged**